server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,433 @@
+            );
+            final Collection<HealthIndicatorImpact> impacts = getUnstableMasterImpacts();
+            Map<String, Object> details;
+            if (explain) {
+                details = new HashMap<>();
+                details.put(DETAILS_CURRENT_MASTER, new DiscoveryNodeXContentObject(localMasterHistory.getMostRecentMaster()));
+                if (remoteHistoryException != null) {
+                    details.put(DETAILS_EXCEPTION_FETCHING_HISTORY, remoteHistoryException.getMessage());
+                    StringWriter stringWriter = new StringWriter();
+                    remoteHistoryException.printStackTrace(new PrintWriter(stringWriter));
+                    String remoteHistoryExceptionStackTrace = stringWriter.toString();
+                    details.put(DETAILS_EXCEPTION_FETCHING_HISTORY_STACK_TRACE, remoteHistoryExceptionStackTrace);
+                }
+            } else {
+                details = Map.of();
+            }
+            final List<UserAction> userActions = getContactSupportUserActions(explain);
+            return createIndicator(
+                stableMasterStatus,
+                summary,
+                explain ? new SimpleHealthIndicatorDetails(details) : HealthIndicatorDetails.EMPTY,
+                impacts,
+                userActions
+            );
+        } else {
+            logger.trace("This node thinks the master is unstable, but the master node {} thinks it is stable", master);
+            return getMasterIsStableResult(explain, localMasterHistory);
+        }
+    }
+
+    /**
+     * Returns a HealthIndicatorResult for the case when the master is seen as stable
+     * @return A HealthIndicatorResult for the case when the master is seen as stable (GREEN status, no impacts or details)
+     */
+    private HealthIndicatorResult getMasterIsStableResult(boolean explain, MasterHistory localMasterHistory) {
+        HealthStatus stableMasterStatus = HealthStatus.GREEN;
+        String summary = "The cluster has a stable master node";
+        Collection<HealthIndicatorImpact> impacts = new ArrayList<>();
+        List<UserAction> userActions = List.of();
+        logger.trace("The cluster has a stable master node");
+        Map<String, Object> details = getSimpleDetailsMap(explain, localMasterHistory);
+        return createIndicator(stableMasterStatus, summary, new SimpleHealthIndicatorDetails(details), impacts, userActions);
+    }
+
+    /**
+     * Returns the health result for the case when we have NOT seen a master recently (at some point in the last 30 seconds).
+     * @param localMasterHistory The master history as seen from the local machine
+     * @param explain Whether to calculate and include the details in the result
+     * @return The HealthIndicatorResult for the given localMasterHistory
+     */
+    private HealthIndicatorResult calculateWhenHaveNotSeenMasterRecently(MasterHistory localMasterHistory, boolean explain) {
```

Comment:
As above, I think this should be named according to what it does/what it looks for as opposed to when it's called (the *when* in the name is confusing as it points to calculating the time when an event happened)

Maybe `onNotSeenMaster` / `onNoLocalMaster` or something similar?